hedera-node/src/main/java/com/hedera/services/state/logic/NetworkCtxManager.java

```diff
@@ -150,17 +158,32 @@ public boolean currentTxnIsFirstInConsensusSecond() {
 		return consensusSecondJustChanged;
 	}
 
-	public void prepareForIncorporating(TxnAccessor accessor) {
-		/* This is only to monitor the current network usage for automated
-		congestion pricing; we don't actually throttle consensus transactions. */
-		handleThrottling.shouldThrottleTxn(accessor);
+	/**
+	 * Used to monitor the current network usage for automated congestion pricing
+	 * and to throttle ContractCreate and ContractCall TXs by the transaction gas limit.
+	 * @param accessor - the accessor for the transaction
+	 * @return - {@link ResponseCodeEnum#OK} if the system has enough capacity to handle the transaction
+	 * {@link ResponseCodeEnum#CONSENSUS_GAS_EXHAUSTED} if the transaction should be throttled
+	 */
+	public ResponseCodeEnum prepareForIncorporating(TxnAccessor accessor) {
+		if(handleThrottling.shouldThrottleConsensusTxn(accessor) &&
+				(accessor.getFunction().equals(ContractCall) || accessor.getFunction().equals(ContractCreate))){
+			return ResponseCodeEnum.CONSENSUS_GAS_EXHAUSTED;
+		}
 
 		feeMultiplierSource.updateMultiplier(networkCtx.get().consensusTimeOfLastHandledTxn());
+		return ResponseCodeEnum.OK;
 	}
 
 	public void finishIncorporating(HederaFunctionality op) {
 		opCounters.countHandled(op);
 
+		if(MiscUtils.isConsensusThrottled(op) &&
+				dynamicProperties.shouldThrottleByGas() &&
+				(txnCtx.status().equals(ResponseCodeEnum.SUCCESS) || txnCtx.status().equals(ResponseCodeEnum.OK))){
```

Comment:
I think that independent of the status of the transaction we might need to leak the unused gas from the reservation